### PR TITLE
Use `insertvalue` and `extractvalue` instead of `memcpy` in CastTarget

### PR DESCRIPTION
**Pull request by dianqk**
r? @nikic

### CODE REVIEW

**Comment by DianQK** [comment]
```
Undefined behavior: Buffer overflow                                      store i64 %2, ptr %3, align 2
```

@rustbot author

File: compiler/rustc_codegen_ssa/src/traits/abi.rs [hunk 1]
```diff
@@ -1,5 +1,148 @@
...
+                    bx.store(from, ptr, align);
+                }
+            }
+            ty_kind if bx.type_kind(bx.reg_backend_type(&self.rest.unit)) == ty_kind => {
+                let scratch_size = self.unaligned_size(bx);
+                let src = if scratch_size > layout.size {
+                    // When using just a single register, we directly use load or store instructions,
+                    // so we must allocate sufficient space.
+                    let scratch_align = self.align(bx);
+                    let llscratch = bx.alloca(scratch_size, scratch_align);
```
Comment (DianQK) [inline compiler/rustc_codegen_ssa/src/traits/abi.rs:134]: This should also apply to other cases.

**Comment by DianQK** [comment]
I believe the PR is far from being ready for review. The main problem is the incorrect implementation of the cast. In addition to using `insertvalue` and `extractvalue`, I also need to handle type conversion here and remove the scratch temporary variable.

I'm not sure whether I should perform a complete conversion like SROA does, or if I should just truncate/convert the final value? Do you have any suggestions? @nikic 

Additionally, I think I could revert the implementation in `compiler/rustc_codegen_ssa/src/traits/abi.rs` to use `memcpy`, and then handle the type conversion in the next PR.

**Comment by Dylan-DPC** [comment]
@DianQK any updates on this? thanks

**Comment by DianQK** [comment]
> @DianQK any updates on this? thanks

I'm low on the priority list for this PR. :)

**Comment by DianQK** [comment]
When I rework on this PR, I'll split it into two separate PRs.

**Pull request closed without merging**